src/compiler/java_compiler.nit

```diff
@@ -0,0 +1,2297 @@
+		if value.is_boxed then return value
+		var obj_type = compiler.mainmodule.object_type
+		if value.mtype isa MNullType then
+			return new_expr("new RTVal(null, null)", compiler.mainmodule.model.null_type)
+		end
+		var mbox = value.mtype.as(MClassType).mclass
+		return new_expr("new RTVal({mbox.rt_name}.get{mbox.rt_name}(), {value})", obj_type)
+	end
+
+	# Unbox primitive `value` to `mtype`.
+	private fun unbox(value: RuntimeVariable, mtype: MType): RuntimeVariable do
+		if not value.is_boxed then return value
+		if not mtype.is_java_primitive then return value
+		if compiler.box_kinds.has(mtype) then
+			return new_expr("({mtype.java_type}){value}.value", mtype)
+		else
+			info "NOT YET IMPLEMENTED unbox for {value} ({mtype})"
+			abort
+		end
+	end
+
+	# Box or unbox primitive `value` to `mtype` if needed.
+	private fun autobox(value: RuntimeVariable, mtype: MType): RuntimeVariable do
+		if mtype.is_java_primitive then return unbox(value, mtype)
+		return box(value, mtype)
+	end
+
+	# Can this `value` be a primitive Java value?
+	private fun can_be_primitive(value: RuntimeVariable): Bool do
+		var t = value.mcasttype.undecorate
+		if not t isa MClassType then return false
+		var k = t.mclass.kind
+		return k == interface_kind or t.is_java_primitive
+	end
+
+	#  Generate a polymorphic subtype test
+	fun type_test(value: RuntimeVariable, mtype: MType): RuntimeVariable do
+		add("/* {value.inspect} isa {mtype} */")
+		var res = self.new_var(compiler.mainmodule.bool_type)
+
+		# check color is in table
+		var maybenull = (value.mcasttype isa MNullableType or value.mcasttype isa MNullType)
+		if maybenull then
+			add("if({value} == null || {value}.is_null()) \{")
+			add("{res} = true && {mtype isa MNullableType};")
+			add("\} else \{")
+		end
+		if mtype isa MNullableType then mtype = mtype.mtype
+		var mclass = mtype.as(MClassType).mclass
+		add("{res} = {value}.rtclass.supers.get(\"{mclass.jname}\") == {mclass.rt_name}.get{mclass.rt_name}();")
+		if maybenull then
+			add("\}")
+		end
+		return res
+	end
+
+	# Generate the code required to dynamically check if 2 objects share the same runtime type
+	fun is_same_type_test(value1, value2: RuntimeVariable): RuntimeVariable do
+		var res = self.new_var(compiler.mainmodule.bool_type)
+		add("{res} = {value1}.rtclass == {value2}.rtclass;")
+		return res
+	end
+
+	# Native instances
+
+	# Generate an integer value
+	fun int_instance(value: Int): RuntimeVariable do
+		var t = compiler.mainmodule.int_type
+		return new RuntimeVariable(value.to_s, t, t)
+	end
+
+	# Generate a byte value
+	fun byte_instance(value: Byte): RuntimeVariable do
+		var t = compiler.mainmodule.byte_type
+		return new RuntimeVariable(value.to_s, t, t)
+	end
+
+	# Generate a char value
+	fun char_instance(value: Char): RuntimeVariable do
+		var t = compiler.mainmodule.char_type
+		return new RuntimeVariable("'{value.to_s.escape_to_c}'", t, t)
+	end
+
+	# Generate a float value
+	#
+	# FIXME pass a Float, not a string
```

Comment:
@xymus the issue is that our float implementation of to_s isn't that reliable and may lose information.